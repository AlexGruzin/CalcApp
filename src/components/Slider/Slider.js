import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import './Slider.styles.less';

const Slider = ({ className, indicator, onChange, defaultValue }) => {
  const [currentValue, setCurrentValue] = useState(defaultValue);

  const handleOnChange = e => {
    onChange(e);
    setCurrentValue(e.target.value);
  };

  const percentHeight = 280 / 100; // TODO 280 - rect overall height get from dom api

  const normalizedPercent = 100 - currentValue;

  return (
    <div className={classNames(className, 'range-slider')}>
      <div className="range-slider__rotator">
        <input
          className="range-slider__range"
          onChange={handleOnChange}
          value={currentValue}
          type="range"
          min="0"
          max="100"
          step="1"
        />
        <span className="slider-bar">
          <span
            className="slider-bar__styled"
            style={{ width: `${percentHeight * normalizedPercent - (16 * normalizedPercent) / 100}px` }}
          />
        </span>
        {indicator && <span className="range-slider__indicator">100</span>}
      </div>
    </div>
  );
};

Slider.propTypes = {
  className: PropTypes.string,
  indicator: PropTypes.string,
  onChange: PropTypes.func,
  defaultValue: PropTypes.number,
};

Slider.defaultProps = {
  className: '',
  defaultValue: 50,
  indicator: false,
  onChange: () => undefined,
};

export default Slider;
