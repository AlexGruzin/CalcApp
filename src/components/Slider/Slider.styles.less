.range-slider {
  display: flex;
  position: relative;
  height: 280px;
  width: 40px;
}

.range-slider__rotator {
  display: flex;
  position: relative;
  height: 1px;
  width: 1px;
  transform: rotate(90deg);
}

.slider-bar {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  position: absolute;
  bottom: 0;
  min-width: 280px;
  height: 40px;
  pointer-events: none;
}

.slider-bar__styled {
  background: green;
  height: 8px;
  z-index: 1;
}

.range-slider input[type=range] {
  -webkit-appearance: none;
  background-color: transparent;
  bottom: 0;
  // border: 1px solid transparent;
  position: absolute;
  // padding: 2px;
  min-width: 280px;
  height: 40px;
}

.range-slider input[type=range]:focus {
  outline: none;
  border: 1px solid var(--outline-default);
}

.range-slider input[type="range"]::-webkit-slider-thumb {
  -webkit-appearance: none;
  z-index: 2;
  height: 36px;
  width: 16px;
  border: 1px solid var(--grey300);
  margin-top: -14px;
  background: var(--grey100);
}

.range-slider input[type=range]::-webkit-slider-runnable-track {
  -webkit-appearance: none;
  background: var(--grey300);
  width: 100%;
  height: 8px;
  cursor: pointer;
}
