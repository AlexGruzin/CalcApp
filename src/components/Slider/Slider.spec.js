import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import Slider from './Slider';

describe('Slider tests', () => {
  it('Check snapshots of Slider are equal.', () => {
    const { asFragment } = render(<Slider onChange={() => undefined} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Check Slider do call onChange on user click.', () => {
    const mockOnnChange = jest.fn();
    const { getByDisplayValue } = render(<Slider onChange={mockOnnChange} />);
    const input = getByDisplayValue('50'); // default range percent

    fireEvent.change(input, { target: { value: '80' } });

    expect(mockOnnChange).toBeCalled();
  });
});
