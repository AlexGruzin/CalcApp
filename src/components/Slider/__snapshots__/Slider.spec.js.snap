// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider tests Check snapshots of Slider are equal. 1`] = `
<DocumentFragment>
  <div
    class="range-slider"
  >
    <div
      class="range-slider__rotator"
    >
      <input
        class="range-slider__range"
        max="100"
        min="0"
        step="1"
        type="range"
        value="50"
      />
      <span
        class="slider-bar"
      >
        <span
          class="slider-bar__styled"
          style="width: 132px;"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
